x/stake/handler.go

```diff
@@ -53,212 +63,245 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
+	k.UpdateValidator(ctx, validator)
 	if err != nil {
 		return err.Result()
 	}
-	tags = tags.AppendTags(delegateTags)
+
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionCreateValidator,
+		tags.DstValidator, msg.ValidatorAddr.Bytes(),
+		tags.Moniker, []byte(msg.Description.Moniker),
+		tags.Identity, []byte(msg.Description.Identity),
+	)
 	return sdk.Result{
 		Tags: tags,
 	}
 }
 
-func handleMsgEditValidator(ctx sdk.Context, msg MsgEditValidator, k Keeper) sdk.Result {
+func handleMsgEditValidator(ctx sdk.Context, msg types.MsgEditValidator, k keeper.Keeper) sdk.Result {
 
 	// validator must already be registered
 	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
 	if !found {
-		return ErrBadValidatorAddr(k.codespace).Result()
-	}
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+		return ErrNoValidatorFound(k.Codespace()).Result()
 	}
 
-	// XXX move to types
 	// replace all editable fields (clients should autofill existing values)
-	validator.Description.Moniker = msg.Description.Moniker
-	validator.Description.Identity = msg.Description.Identity
-	validator.Description.Website = msg.Description.Website
-	validator.Description.Details = msg.Description.Details
+	description, err := validator.Description.UpdateDescription(msg.Description)
+	if err != nil {
+		return err.Result()
+	}
+	validator.Description = description
 
-	k.updateValidator(ctx, validator)
+	k.UpdateValidator(ctx, validator)
 	tags := sdk.NewTags(
-		"action", []byte("editValidator"),
-		"validator", msg.ValidatorAddr.Bytes(),
-		"moniker", []byte(msg.Description.Moniker),
-		"identity", []byte(msg.Description.Identity),
+		tags.Action, tags.ActionEditValidator,
+		tags.DstValidator, msg.ValidatorAddr.Bytes(),
+		tags.Moniker, []byte(description.Moniker),
+		tags.Identity, []byte(description.Identity),
 	)
 	return sdk.Result{
 		Tags: tags,
 	}
 }
 
-func handleMsgDelegate(ctx sdk.Context, msg MsgDelegate, k Keeper) sdk.Result {
+func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper) sdk.Result {
 
 	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
 	if !found {
-		return ErrBadValidatorAddr(k.codespace).Result()
+		return ErrNoValidatorFound(k.Codespace()).Result()
 	}
 	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
-		return ErrBadBondingDenom(k.codespace).Result()
+		return ErrBadDenom(k.Codespace()).Result()
 	}
 	if validator.Revoked == true {
-		return ErrValidatorRevoked(k.codespace).Result()
+		return ErrValidatorRevoked(k.Codespace()).Result()
 	}
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
-	}
-	tags, err := delegate(ctx, k, msg.DelegatorAddr, msg.Bond, validator)
+	_, delegation, validator, pool, err := k.Delegate(ctx, msg.DelegatorAddr, msg.Bond, validator)
 	if err != nil {
 		return err.Result()
 	}
+	k.SetPool(ctx, pool)
+	k.SetDelegation(ctx, delegation)
+	k.UpdateValidator(ctx, validator)
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionDelegate,
+		tags.Delegator, msg.DelegatorAddr.Bytes(),
```

Comment:
Ditto as above (address encoding in tags) for this line and the below line

Reply:
addressed